batcher/aligned-batcher/src/types/batch_queue.rs

```diff
@@ -87,9 +110,428 @@ impl Ord for BatchQueueEntryPriority {
         if ord == std::cmp::Ordering::Equal {
             self.nonce.cmp(&other.nonce).reverse()
         } else {
-            ord
+            ord.reverse()
         }
     }
 }
 
 pub(crate) type BatchQueue = PriorityQueue<BatchQueueEntry, BatchQueueEntryPriority>;
+
+/// Calculates the size of the batch represented by the given batch queue.
+pub(crate) fn calculate_batch_size(batch_queue: &BatchQueue) -> Result<usize, BatcherError> {
+    let folded_result = batch_queue.iter().try_fold(0, |acc, (entry, _)| {
+        if let Ok(verification_data_bytes) =
+            cbor_serialize(&entry.nonced_verification_data.verification_data)
+        {
+            let current_batch_size = acc + verification_data_bytes.len();
+            ControlFlow::Continue(current_batch_size)
+        } else {
+            ControlFlow::Break(())
+        }
+    });
+
+    if let ControlFlow::Continue(batch_size) = folded_result {
+        Ok(batch_size)
+    } else {
+        Err(BatcherError::SerializationError(String::from(
+            "Could not calculate size of batch",
+        )))
+    }
+}
+
+/// This function tries to build a batch to be submitted to Aligned.
+/// Given a copy of the current batch queue, , and applyies an algorithm to find the biggest batch
+/// of proofs from users that are willing to pay for it:
+/// 1. Traverse each batch priority queue, starting from the one with minimum max fee.
+/// 2. Calculate the `fee_per_proof` for the whole batch and compare with the `max_fee` of the entry.
+/// 3. If `fee_per_proof` is less than the `max_fee` of the current entry, submit the batch. If not, pop this entry
+///     from the queue and push it to `resulting_priority_queue`, then repeat step 1.
+///
+/// `resulting_priority_queue` will be the batch queue composed of all entries that were not willing to pay for the batch.
+/// This is outputted in along with the finalized batch.
+pub(crate) fn try_build_batch(
+    batch_queue_copy: &mut BatchQueue,
```

Comment:
If it needs to be a copy, either don't use a reference, or use a non mutable reference and clone inside the function, to avoid mistakes

Reply:
solved in 31c9ae2877f7746a868e029fb0a2c50df68181d6